source/crud/tests/README.rst

```diff
@@ -90,10 +78,41 @@ Each YAML file has the following keys:
       - ``data``: The data that should exist in the collection after the
         operation has been run.
 
+Legacy Test Format for Single Operations
+----------------------------------------
+
+The test format above supports both multiple operations and APM expectations,
+and is consistent with the formats used by other specifications. Previously, the
+CRUD spec tests used a simplified format that only allowed for executing a
+single operation. Notable differences from the current format are as follows:
+
+- Instead of a `tests[i].operations` array, a single operation was defined as a
+  document in `tests[i].operation`. That document consisted of only the `name`,
+  `arguments`, and an optional `object` field.
+
+- Instead of `error` and `result` fields within each element in the
+  `tests[i].operations` array, the single operation's error and result were
+  defined under the `tests[i].outcome.error` and `tests[i].outcome.result`
+  fields.
+
+The legacy format should not conflict with the newer, multi-operation format
+used by other specs. Several drivers currently handle both formats using a
+unified test runner.
+
 Expectations
 ============
 
+Optional Fields in Results
+--------------------------
+
 Expected results for some tests may include optional fields, such as
 ``insertedId`` (for InsertOneResult), ``insertedIds`` (for InsertManyResult),
 and ``upsertedCount`` (for UpdateResult). Drivers that do not implement these
 fields can ignore them.
+
+Asserting Nonexistent Fields
```

Comment:
This is something I've had in the back of my mind for a while. On a related note, I think we can use `{ $exists: true }` in place of magic number `42` in other spec tests (e.g. checking for the presence of values we don't particular care about, like cursor IDs).

Thoughts? If this is controversial, I can revert to the [Null Values](https://github.com/mongodb/specifications/tree/master/source/transactions/tests#null-values) language from the Transactions spec.

Reply:
I do prefer this to `null`/`42` (even though I am a big H2G2 fan) as it makes the test writer's intentions explicit.

Reply:
I do prefer this to `null`/`42` (even though I am a big H2G2 fan) as it makes the test writer's intentions explicit.

Reply:
Added sections to explain how `{ $exists: true }` and `{ $exists: false }` work. That said, these still aren't necessary for any existing CRUD tests. My goal here is to get this accepted so we can start using it going forward (and ultimately replace the less explicit behavior in other spec runners).

Reply:
I am a big fan of this idea in general, but since [`$exists` is a valid query operator](https://docs.mongodb.com/manual/reference/operator/query/exists/), could we use something that would never be used in MQL?

Reply:
Just out of curiosity, why is it problematic to have a construct that _may_ also be a part of MQL in an object containing *expected* key, value pairs? AFAICT, the expectations will only ever be parsed by the language implementing the spec and never by the server...

Reply:
@prashantmital I am worried about introspection of outgoing queries. For example, with the following APM event:

```
{
  "command": {
    "find": "bar",
    "filter": {
      "elbow": {
        "$exists": false
      }
    }
    ...
  }
  ...
}
```

How would we write the `expected` version of that event?

Reply:
Ah, very good point. I really did not expect `$exists` to be used anywhere other than the top-level of a document -- although I didn't explicitly mention that in the README. Re-reading the [match documentation](https://github.com/mongodb/specifications/tree/master/source/change-streams/tests#spec-test-match-function) from the Change Stream spec, it seems like you intended assertions of `42` to occur at any level of the document, but I wonder if that's actually ever needed in practice. In the Transactions spec, I believe assertions for nonexistent fields via `null` is only done at the top-level (e.g. absence of a `writeConcern` field).

If we apply that restriction, I don't think we'd have a risk of conflict with actual query operators. A real `$exists` operator might appear under a `filter` key for a `find` command (or similar, such as `query` for `findAndModify`) or deeper withing an aggregation pipeline.

If you do think it's important to allow this comparison at any level, we can create a more contrived name. In my experience with PHP, at some point in our test runner I just convert the document to extended JSON and do a straight up comparison rather than crawl the document recursively (as suggested in the Change Stream spec test algorithm). I'm not sure what other drivers are doing.

Reply:
How about `{"$assertExists": true}`?

Reply:
Created [SPEC-1215](https://jira.mongodb.org/browse/SPEC-1215) to track this idea. I'll remove both of these sections from the PR, as the original text is now quoted in that ticket.